In [ ]:
# Initialize Otter
import otter
grader = otter.Notebook("lab03.ipynb")

# Lab 03: Data Cleaning and EDA
## Due Wednesday, June 26th, 11:59 PM PT

To receive credit for a lab, answer all questions correctly and submit before the deadline.

You must submit this assignment to Gradescope by the on-time deadline, **Wednesday, June 26th, 11:59 PM PT**. Please read the syllabus for the grace period policy. No late submissions beyond the grace period will be accepted. While course staff is happy to help you if you encounter difficulties with submission, we may not be able to respond to late-night requests for assistance (TAs need to sleep, after all!). **We strongly encourage you to plan to submit your work to Gradescope several hours before the stated deadline.** This way, you will have ample time to contact staff for submission support. 

### Lab Walk-Through
In addition to the lab notebook, we have also released a prerecorded walk-through video of the lab. We encourage you to reference this video as you work through the lab. Run the cell below to display the videos.
<br>
**Note:** This video was recorded in Fall 2023. **Question 5 is a new addition this summer and will not be included in the video**. Please reach out for help in office hours or on Ed if you have any questions regarding this portion of the lab! 

In [ ]:
from IPython.display import YouTubeVideo
YouTubeVideo("hwRYs5ZRgW4", list = 'PLQCcNQgUcDfqSg049DVFZCQbupMY5Bn5Z', listType = 'playlist')

### Collaboration Policy

Data science is a collaborative activity. While you may talk with others about the labs, we ask that you **write your solutions individually**. If you do discuss the assignments with others, please **include their names** below. (That's a good way to learn your classmates' names.)

**Collaborators**: *list collaborators here*

---
### Debugging Guide
If you run into any technical issues, we highly recommend checking out the [Data 100 Debugging Guide](https://ds100.org/debugging-guide/). In this guide, you can find general questions about Jupyter notebooks / Datahub, Gradescope, and common pandas errors.

---
[`pandas`](https://pandas.pydata.org/) is one of the most widely used `Python` libraries in data science. In this lab, you will review commonly used data-wrangling operations/tools in `pandas`. We continue the content from the previous lab and aim to give you familiarity with:

* Aggregating the data (using `.groupby`),
* Filtering the data (using boolean arrays and `groupby.filter`),
* Pivoting (using `.pivot_table`).

In this lab, you are going to use several `pandas` methods. Reminder from the lecture that you may press `shift+tab` on method parameters to see the documentation for that method. For example, if you were using the `drop` method in `pandas`, you could press `shift+tab` to see what `drop` is expecting.

`pandas` is very similar to the `datascience` library that you saw in Data 8. This [conversion notebook](https://data100.datahub.berkeley.edu/hub/user-redirect/git-pull?repo=https%3A%2F%2Fgithub.com%2FDS-100%2Fsu23-materials&branch=main&urlpath=lab%2Ftree%2Fsu23-materials%2Flec%2Flec02%2Fdata8_translation_examples.ipynb) may serve as a useful guide!

This lab expects that you have watched all three `pandas` lectures. If you have not, this lab will probably take a very long time.

In [72]:
import numpy as np
import pandas as pd
%matplotlib inline

<hr style="border: 5px solid #003262;" />
<hr style="border: 1px solid #fdb515;" />

### **REVIEW:** `Groupby` and `Groupby` Shorthand

Let's now turn to use `groupby` from lectures 3 and 4.

### Elections

Let's start by reading in the election dataset from the `pandas` lectures.

In [73]:
# Run this cell to load data from CSV file; no further action is needed.
elections = pd.read_csv("data/elections.csv")
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

As we saw before, we can `groupby` a specific column, e.g., `"Party"` and can print out the resulting sub-DataFrames. The output below can help you get an understanding of what `groupby` is actually doing.

An example is given below for elections since 1980.

In [74]:
# Run this cell to print sub-DataFrames of a groupby object; no further action is needed.
for n, g in elections[elections["Year"] >= 1980].groupby("Party"):
    print(f"Name: {n}") # By the way, this is an "f string", a relatively new and great feature of Python
    display(g)

Name: Citizens


Year       Candidate     Party  Popular vote Result         %
127  1980  Barry Commoner  Citizens        233052   loss  0.270182

Name: Constitution


Year         Candidate         Party  Popular vote Result         %
160  2004  Michael Peroutka  Constitution        143630   loss  0.117542
164  2008     Chuck Baldwin  Constitution        199750   loss  0.152398
172  2016    Darrell Castle  Constitution        203091   loss  0.149640

Name: Democratic


Year        Candidate       Party  Popular vote Result          %
129  1980     Jimmy Carter  Democratic      35480115   loss  41.132848
134  1984   Walter Mondale  Democratic      37577352   loss  40.729429
137  1988  Michael Dukakis  Democratic      41809074   loss  45.770691
140  1992     Bill Clinton  Democratic      44909806    win  43.118485
144  1996     Bill Clinton  Democratic      47400125    win  49.296938
151  2000          Al Gore  Democratic      50999897   loss  48.491813
158  2004       John Kerry  Democratic      59028444   loss  48.306775
162  2008     Barack Obama  Democratic      69498516    win  53.023510
168  2012     Barack Obama  Democratic      65915795    win  51.258484
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515

Name: Green


Year         Candidate  Party  Popular vote Result         %
149  1996       Ralph Nader  Green        685297   loss  0.712721
155  2000       Ralph Nader  Green       2882955   loss  2.741176
156  2004        David Cobb  Green        119859   loss  0.098088
165  2008  Cynthia McKinney  Green        161797   loss  0.123442
170  2012        Jill Stein  Green        469627   loss  0.365199
177  2016        Jill Stein  Green       1457226   loss  1.073699
181  2020    Howard Hawkins  Green        405035   loss  0.255731

Name: Independent


Year         Candidate        Party  Popular vote Result          %
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

Name: Libertarian


Year         Candidate        Party  Popular vote Result         %
128  1980          Ed Clark  Libertarian        921128   loss  1.067883
132  1984    David Bergland  Libertarian        228111   loss  0.247245
138  1988          Ron Paul  Libertarian        431750   loss  0.472660
139  1992      Andre Marrou  Libertarian        290087   loss  0.278516
146  1996      Harry Browne  Libertarian        485759   loss  0.505198
153  2000      Harry Browne  Libertarian        384431   loss  0.365525
159  2004  Michael Badnarik  Libertarian        397265   loss  0.325108
163  2008          Bob Barr  Libertarian        523715   loss  0.399565
169  2012      Gary Johnson  Libertarian       1275971   loss  0.992241
175  2016      Gary Johnson  Libertarian       4489235   loss  3.307714
180  2020      Jo Jorgensen  Libertarian       1865724   loss  1.177979

Name: Natural Law


Year     Candidate        Party  Popular vote Result         %
148  1996  John Hagelin  Natural Law        113670   loss  0.118219

Name: New Alliance


Year      Candidate         Party  Popular vote Result         %
136  1988  Lenora Fulani  New Alliance        217221   loss  0.237804

Name: Populist


Year Candidate     Party  Popular vote Result         %
141  1992  Bo Gritz  Populist        106152   loss  0.101918

Name: Reform


Year     Candidate   Party  Popular vote Result         %
150  1996    Ross Perot  Reform       8085294   loss  8.408844
154  2000  Pat Buchanan  Reform        448895   loss  0.426819

Name: Republican


Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
142  1992  George H. W. Bush  Republican      39104550   loss  37.544784
145  1996           Bob Dole  Republican      39197469   loss  40.766036
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
166  2008        John McCain  Republican      59948323   loss  45.737243
171  2012        Mitt Romney  Republican      60933504   loss  47.384076
173  2016       Donald Trump  Republican      62984828    win  46.407862
179  2020       Donald Trump  Republican      74216154   loss  46.858542

Name: Taxpayers


Year        Candidate      Party  Popular vote Result         %
147  1996  Howard Phillips  Taxpayers        184656   loss  0.192045

Recall that once we've formed groups, we can aggregate each sub-DataFrame (a.k.a. group) into a single row using an aggregation function. For example, if we use `.agg('mean')` on the groups above, we get back a single `DataFrame` where each group has been replaced by a single row. In each column for that aggregate row, the value that appears is the average of all values in that group.

For columns that are non-numeric, e.g., `"Result"`, the `pandas` version we're using (version 2.0.2) will error because we cannot compute the mean of the `Result` column. To remedy this, we add a `numeric_only=True` argument to our function calls so that we only calculate the `mean` for columns that contain numeric values. Alternatively, we can manually select only the numerical columns before calling the `agg` function so the aggregation is only applied to numerical columns.

In [75]:
elections_after_1980 = elections[elections["Year"] >= 1980]

elections_after_1980.groupby("Party").agg('mean', numeric_only=True)

# alternatively, we can manually select only the numerical columns before calling `agg`
# elections_after_1980.groupby("Party")[['Year', 'Popular vote', '%']].agg('mean')

Year  Popular vote          %
Party                                             
Citizens      1980.000000  2.330520e+05   0.270182
Constitution  2009.333333  1.821570e+05   0.139860
Democratic    2000.000000  5.452196e+07  47.360184
Green         2008.000000  8.831137e+05   0.767151
Independent   2000.000000  5.480026e+06   5.414298
Libertarian   2000.000000  1.026652e+06   0.830876
Natural Law   1996.000000  1.136700e+05   0.118219
New Alliance  1988.000000  2.172210e+05   0.237804
Populist      1992.000000  1.061520e+05   0.101918
Reform        1998.000000  4.267094e+06   4.417831
Republican    2000.000000  5.419334e+07  47.898650
Taxpayers     1996.000000  1.846560e+05   0.192045

Equivalently we can use one of the shorthand aggregation functions, e.g. `.mean()`: 

In [76]:
elections_after_1980.groupby("Party").mean(numeric_only=True)

Year  Popular vote          %
Party                                             
Citizens      1980.000000  2.330520e+05   0.270182
Constitution  2009.333333  1.821570e+05   0.139860
Democratic    2000.000000  5.452196e+07  47.360184
Green         2008.000000  8.831137e+05   0.767151
Independent   2000.000000  5.480026e+06   5.414298
Libertarian   2000.000000  1.026652e+06   0.830876
Natural Law   1996.000000  1.136700e+05   0.118219
New Alliance  1988.000000  2.172210e+05   0.237804
Populist      1992.000000  1.061520e+05   0.101918
Reform        1998.000000  4.267094e+06   4.417831
Republican    2000.000000  5.419334e+07  47.898650
Taxpayers     1996.000000  1.846560e+05   0.192045

Note that the index of the `DataFrame` returned by an `groupby.agg` call is no longer a set of numeric indices from $0$ to $N-1$. Instead, we see that the index for the example above is now the `Party`. If we want to restore our `DataFrame` so that `Party` is a column rather than the index, we can use `reset_index`.

In [77]:
elections_after_1980.groupby("Party").mean(numeric_only=True).reset_index()

Party         Year  Popular vote          %
0       Citizens  1980.000000  2.330520e+05   0.270182
1   Constitution  2009.333333  1.821570e+05   0.139860
2     Democratic  2000.000000  5.452196e+07  47.360184
3          Green  2008.000000  8.831137e+05   0.767151
4    Independent  2000.000000  5.480026e+06   5.414298
5    Libertarian  2000.000000  1.026652e+06   0.830876
6    Natural Law  1996.000000  1.136700e+05   0.118219
7   New Alliance  1988.000000  2.172210e+05   0.237804
8       Populist  1992.000000  1.061520e+05   0.101918
9         Reform  1998.000000  4.267094e+06   4.417831
10    Republican  2000.000000  5.419334e+07  47.898650
11     Taxpayers  1996.000000  1.846560e+05   0.192045

**IMPORTANT NOTE:** Notice that the code above consists of chained method calls. This sort of code is very common in `pandas` programming and in data science in general. Such chained method calls can sometimes go many layers deep, so you might consider adding newlines between lines of code for clarity. For example, we could instead write the code above as:

In [78]:
# pandas method chaining
(
elections.query("Year >= 1980").groupby("Party") 
                               .mean(numeric_only=True)  ## Computes the mean values by party
                               .reset_index()            ## Resets to a numerical index
)

Party         Year  Popular vote          %
0       Citizens  1980.000000  2.330520e+05   0.270182
1   Constitution  2009.333333  1.821570e+05   0.139860
2     Democratic  2000.000000  5.452196e+07  47.360184
3          Green  2008.000000  8.831137e+05   0.767151
4    Independent  2000.000000  5.480026e+06   5.414298
5    Libertarian  2000.000000  1.026652e+06   0.830876
6    Natural Law  1996.000000  1.136700e+05   0.118219
7   New Alliance  1988.000000  2.172210e+05   0.237804
8       Populist  1992.000000  1.061520e+05   0.101918
9         Reform  1998.000000  4.267094e+06   4.417831
10    Republican  2000.000000  5.419334e+07  47.898650
11     Taxpayers  1996.000000  1.846560e+05   0.192045

Note that we have surrounded the entire call by a big set of parentheses so that `Python` doesn't complain about the indentation. An alternative is to use the \ symbol to indicate to `Python` that your code continues on to the next line!

In [79]:
# pandas method chaining (alternative)
elections[elections["Year"] >= 1980].groupby("Party") \
                               .mean(numeric_only=True) \
                               .reset_index()     

Party         Year  Popular vote          %
0       Citizens  1980.000000  2.330520e+05   0.270182
1   Constitution  2009.333333  1.821570e+05   0.139860
2     Democratic  2000.000000  5.452196e+07  47.360184
3          Green  2008.000000  8.831137e+05   0.767151
4    Independent  2000.000000  5.480026e+06   5.414298
5    Libertarian  2000.000000  1.026652e+06   0.830876
6    Natural Law  1996.000000  1.136700e+05   0.118219
7   New Alliance  1988.000000  2.172210e+05   0.237804
8       Populist  1992.000000  1.061520e+05   0.101918
9         Reform  1998.000000  4.267094e+06   4.417831
10    Republican  2000.000000  5.419334e+07  47.898650
11     Taxpayers  1996.000000  1.846560e+05   0.192045

**IMPORTANT NOTE:** You should NEVER solve problems like the one above using loops or list comprehensions. This is slow and also misses the entire point of this part of Data 100. 

Before we continue, we'll print out the election dataset again for your convenience. 

In [80]:
elections.head(5)

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789

<br>

---

### Question 1a
Using `groupby.agg` or one of the shorthand methods (`groupby.min`, `groupby.first`, etc.), create a `Series` object `best_result_percentage_only` that returns a `Series` showing the entire best result for every party, sorted in decreasing order. Your `Series` should include only parties that have earned at least 10% of the vote in some election. Your result should look like this:

<code>
Party
Democratic               61.344703
Republican               60.907806
Democratic-Republican    57.210122
National Union           54.951512
Whig                     53.051213
Liberal Republican       44.071406
National Republican      43.796073
Northern Democratic      29.522311
Progressive              27.457433
American                 21.554001
Independent              18.956298
Southern Democratic      18.138998
American Independent     13.571218
Constitutional Union     12.639283
Free Soil                10.138474
Name: %, dtype: float64
</code>
<br/>

A list of named `groupby.agg` shorthand methods are [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#aggregation) (you'll have to scroll down about one page).


In [81]:
best_result_percentage_only = elections.groupby("Party").agg('max', numeric_only=True)
best_result_percentage_only = best_result_percentage_only[best_result_percentage_only['%'] >= 10]
best_result_percentage_only = best_result_percentage_only['%'].sort_values(ascending=False)
best_result_percentage_only

Party
Democratic               61.344703
Republican               60.907806
Democratic-Republican    57.210122
National Union           54.951512
Whig                     53.051213
Liberal Republican       44.071406
National Republican      43.796073
Northern Democratic      29.522311
Progressive              27.457433
American                 21.554001
Independent              18.956298
Southern Democratic      18.138998
American Independent     13.571218
Constitutional Union     12.639283
Free Soil                10.138474
Name: %, dtype: float64

In [82]:
grader.check("q1a")

q1a results:
    q1a - 1 result:
        ✅ Test case passed

    q1a - 2 result:
        ❌ Test case failed
        Trying:
            best_result_percentage_only['Independent'].sum()
        Expecting:
            18.95629754
        **********************************************************************
        Line 1, in q1a 1
        Failed example:
            best_result_percentage_only['Independent'].sum()
        Expected:
            18.95629754
        Got:
            np.float64(18.95629754)

    q1a - 3 result:
        ❌ Test case failed
        Trying:
            best_result_percentage_only.iloc[0]
        Expecting:
            61.34470329
        **********************************************************************
        Line 1, in q1a 2
        Failed example:
            best_result_percentage_only.iloc[0]
        Expected:
            61.34470329
        Got:
            np.float64(61.34470329)

<br>

---

### Question 1b  
Repeat Question 1a. However, this time, your result should be a `DataFrame` showing all available information (all columns) rather than only the percentage as a `Series`.

This question is trickier than Question 1a. Make sure to check the `pandas` lecture slides if you're stuck! It's very easy to make a subtle mistake that shows Woodrow Wilson and Howard Taft both winning the 2020 election.

For example, the first 3 rows of your table should be:

|Party | Year | Candidate      | Popular Vote | Result | %         |
|------|------|----------------|--------------|--------|-----------|
|**Democratic**  | 1964 | Lyndon Johnson | 43127041      | win   | 61.344703 |
|**Republican**  | 1972 | Richard Nixon | 47168710      | win   | 60.907806 |
|**Democratic-Republican**  | 1824 | Andrew Jackson | 151271      | loss   | 57.210122 |

Note that the index is `Party`. In other words, don't use `reset_index`.


In [83]:
best_result = elections.sort_values(by = '%', ascending = False)
best_result = best_result[best_result['%'] >= 10].groupby("Party").first().sort_values(by = '%', ascending = False)
best_result

Year               Candidate  Popular vote Result  \
Party                                                                      
Democratic             1964          Lyndon Johnson      43127041    win   
Republican             1972           Richard Nixon      47168710    win   
Democratic-Republican  1824          Andrew Jackson        151271   loss   
National Union         1864         Abraham Lincoln       2211317    win   
Whig                   1840  William Henry Harrison       1275583    win   
Liberal Republican     1872          Horace Greeley       2834761   loss   
National Republican    1828       John Quincy Adams        500897   loss   
Northern Democratic    1860      Stephen A. Douglas       1380202   loss   
Progressive            1912      Theodore Roosevelt       4122721   loss   
American               1856        Millard Fillmore        873053   loss   
Independent            1992              Ross Perot      19743821   loss   
Southern Democratic    1860    John C. Breckinridge        848019   loss   
American Independent   1968          George Wallace       9901118   loss   
Constitutional Union   1860               John Bell        590901   loss   
Free Soil              1848        Martin Van Buren        291501   loss   

                               %  
Party                             
Democratic             61.344703  
Republican             60.907806  
Democratic-Republican  57.210122  
National Union         54.951512  
Whig                   53.051213  
Liberal Republican     44.071406  
National Republican    43.796073  
Northern Democratic    29.522311  
Progressive            27.457433  
American               21.554001  
Independent            18.956298  
Southern Democratic    18.138998  
American Independent   13.571218  
Constitutional Union   12.639283  
Free Soil              10.138474

In [84]:
grader.check("q1b")

q1b results:
    q1b - 1 result:
        ✅ Test case passed

    q1b - 2 result:
        ❌ Test case failed
        Trying:
            best_result['Popular vote'].sum()
        Expecting:
            135020916
        **********************************************************************
        Line 1, in q1b 1
        Failed example:
            best_result['Popular vote'].sum()
        Expected:
            135020916
        Got:
            np.int64(135020916)

    q1b - 3 result:
        ✅ Test case passed

### **REVIEW:** `DataFrameGroupBy.filter`

Our `DataFrame` contains a number of parties that have never had a successful presidential run. For example, the 2020 elections included candidates from the Libertarian and Green parties, neither of which have elected a president.

In [85]:
# Run this cell to print the last four rows; no further action is needed.
elections.tail(4)

Year       Candidate        Party  Popular vote Result          %
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

Suppose we were conducting an analysis trying to focus our attention on parties that had elected a president. 

The most natural approach is to use `groupby.filter` [(docs)](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.filter.html). This is an incredibly powerful but subtle tool for filtering data.

The code below accomplishes the task at hand. It does this by creating a function that returns `True` if and only if a sub-`DataFrame` (a.k.a. group) contains at least one winner. This function, in turn, uses the `pandas` function `any` [(docs)](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).

In [86]:
# Run this cell to keep only the rows of parties that have 
# elected a president; no further action is needed.
def at_least_one_candidate_in_the_frame_has_won(frame):
    """Returns df with rows only kept for parties that have
    won at least one election
    """
    return (frame["Result"] == 'win').any()

winners_only = (
    elections
        .groupby("Party")
        .filter(at_least_one_candidate_in_the_frame_has_won)
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542

Alternately, we could have used a `lambda` function instead of explicitly defining a named function using `def`. 

In [87]:
# Run this cell to keep only the rows of parties that have 
# elected a president; no further action is needed.
winners_only = (
    elections
        .groupby("Party")
        .filter(lambda x : (x["Result"] == "win").any())
)
winners_only.tail(5)

Year        Candidate       Party  Popular vote Result          %
171  2012      Mitt Romney  Republican      60933504   loss  47.384076
173  2016     Donald Trump  Republican      62984828    win  46.407862
176  2016  Hillary Clinton  Democratic      65853514   loss  48.521539
178  2020     Joseph Biden  Democratic      81268924    win  51.311515
179  2020     Donald Trump  Republican      74216154   loss  46.858542

<br>

---

### Question 1c

Using `filter`, create a `DataFrame` object `major_party_results_since_1988` that includes all election results starting after 1988 (exclusive) but only shows a row if the Party it belongs to has earned at least 1% of the popular vote in ANY election since 1988.

For example, in 1988, you should not include the `New Alliance` candidate since this party has not earned 1% of the vote since 1988. However, you should include the `Libertarian` candidate from 1988 despite only having 0.47 percent of the vote in 1988, because in 2016 and 2020, the Libertarian candidates Gary Johnson and Jo Jorgensen exceeded 1% of the vote.

For example, the first three rows of the table you generate should look like:

|     |   Year | Candidate         | Party       |   Popular vote | Result   |         % |
|----:|-------:|:------------------|:------------|---------------:|:---------|----------:|
| 139 |   1992 | Andre Marrou      | Libertarian |       290087   | loss     | 0.278516  |
| 140 |   1992 | Bill Clinton      | Democratic  |       44909806 | win      | 43.118485 |
| 142 |   1992 | George H. W. Bush | Republican  |       39104550 | loss     |  37.544784|

*Hint*: The following questions might help you construct your solution. One of the lines should be identical to the `filter` examples shown above.

1) How can we **only** keep rows in the data starting after 1988 (exclusive)?
2) What column should we `groupby` to filter out parties that have earned at least 1% of the popular vote in ANY election since 1988?
3) How can we write an aggregation function that takes a sub-DataFrame and returns whether at least 1% of the vote has been earned in that sub-DataFrame? This may give you a hint about the second question!


In [88]:
results_since_1988 = elections[elections['Year'] > 1988]
major_party = results_since_1988.groupby("Party").filter(lambda x: (x['%'] >= 1).any())['Party'].unique()
# display(major_party)
major_party_results_since_1988 = results_since_1988[results_since_1988['Party'].isin(major_party)]
major_party_results_since_1988.head()

Year          Candidate        Party  Popular vote Result          %
139  1992       Andre Marrou  Libertarian        290087   loss   0.278516
140  1992       Bill Clinton   Democratic      44909806    win  43.118485
142  1992  George H. W. Bush   Republican      39104550   loss  37.544784
143  1992         Ross Perot  Independent      19743821   loss  18.956298
144  1996       Bill Clinton   Democratic      47400125    win  49.296938

In [89]:
grader.check("q1c")

q1c results:
    q1c - 1 result:
        ✅ Test case passed

    q1c - 2 result:
        ❌ Test case failed
        Trying:
            major_party_results_since_1988['%'].min()
        Expecting:
            0.098088334
        **********************************************************************
        Line 1, in q1c 1
        Failed example:
            major_party_results_since_1988['%'].min()
        Expected:
            0.098088334
        Got:
            np.float64(0.098088334)

    q1c - 3 result:
        ✅ Test case passed

### **REVIEW:** `str`

`pandas` provides special purpose functions for working with specific common data types such as strings and dates, which you will learn about in more detail in Lecture 6. For example, the code below provides the length of every Candidate's name from our `elections` dataset. 

In [90]:
elections["Candidate"].str.len()

0      14
1      17
2      14
3      17
4      14
       ..
177    10
178    12
179    12
180    12
181    14
Name: Candidate, Length: 182, dtype: int64

<br>

---

### Question 2

Using `.str.split`, create a new `DataFrame` called `elections_with_first_name` with a new column `First Name` that is equal to the Candidate's first name.

See the `pandas` `str` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.split.html) for documentation on using `str.split`.

In [103]:
elections_with_first_name = elections.copy()
elections_with_first_name["First Name"] = elections["Candidate"].str.split(expand=True)[0]
elections_with_first_name
# display(elections_with_first_name[elections_with_first_name["Party"] == 'American'])

Year           Candidate     Party  Popular vote Result          %  \
22   1856    Millard Fillmore  American        873053   loss  21.554001   
126  1976  Thomas J. Anderson  American        158271   loss   0.194862   

    First Name  
22     Millard  
126     Thomas

In [92]:
grader.check("q2")

q2 results: All test cases passed!

<br>

---

## Babynames
Remember the `babynames` dataset from Lab02A? Let's load it in again and explore the data with our newly covered functions! Like last time, we'll only load in data from California. 

Run the following cell: 

In [93]:
file_path = 'data/namesbystate_ca.txt.gz'
column_labels = ['State', 'Sex', 'Year', 'Name', 'Count']

babynames = pd.read_csv(file_path, names=column_labels)

babynames.head()

State Sex  Year      Name  Count
0    CA   F  1910      Mary    295
1    CA   F  1910     Helen    239
2    CA   F  1910   Dorothy    220
3    CA   F  1910  Margaret    163
4    CA   F  1910   Frances    134

The code below creates a table with the frequency of all names from 2022. 

In [94]:
# Run this cell to create a table with the frequency 
# of all names from 2022; no further action is needed.
babynames_2022 = (
    babynames[babynames['Year'] == 2022]
              .groupby("Name")
              .sum()[["Count"]]
              .reset_index()
)
babynames_2022

Name  Count
0     Aadhini      6
1     Aadhira      5
2      Aadhya     33
3        Aadi     11
4       Aadit      5
...       ...    ...
6196     Zyla     20
6197    Zylah     14
6198     Zylo      5
6199     Zyon     17
6200     Zyra     16

[6201 rows x 2 columns]

<br>

---

### Question 3

Using the [pd.merge](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html) function described in the lecture, combine the `babynames_2022` table with the `elections_with_first_name` table you created earlier to form `presidential_candidates_and_name_popularity`. Your resulting `DataFrame` should contain all columns from both of the tables.

In [95]:
presidential_candidates_and_name_popularity = pd.merge(elections_with_first_name, babynames_2022, left_on = 'First Name', right_on = 'Name', how = 'inner')
presidential_candidates_and_name_popularity

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
139  2016       Gary Johnson            Libertarian       4489235   loss   
140  2016    Hillary Clinton             Democratic      65853514   loss   
141  2020       Joseph Biden             Democratic      81268924    win   
142  2020       Donald Trump             Republican      74216154   loss   
143  2020     Howard Hawkins                  Green        405035   loss   

             % First Name     Name  Count  
0    57.210122     Andrew   Andrew    741  
1    42.789878       John     John    490  
2    56.203927     Andrew   Andrew    741  
3    43.796073       John     John    490  
4    54.574789     Andrew   Andrew    741  
..         ...        ...      ...    ...  
139   3.307714       Gary     Gary     19  
140  48.521539    Hillary  Hillary     10  
141  51.311515     Joseph   Joseph    785  
142  46.858542     Donald   Donald     33  
143   0.255731     Howard   Howard     18  

[144 rows x 9 columns]

In [96]:
grader.check("q3")

q3 results:
    q3 - 1 result:
        ✅ Test case passed

    q3 - 2 result:
        ✅ Test case passed

    q3 - 3 result:
        ❌ Test case failed
        Trying:
            presidential_candidates_and_name_popularity[presidential_candidates_and_name_popularity['Candidate'] == 'William Lemke'].iloc[0]['Count']
        Expecting:
            831
        **********************************************************************
        Line 1, in q3 2
        Failed example:
            presidential_candidates_and_name_popularity[presidential_candidates_and_name_popularity['Candidate'] == 'William Lemke'].iloc[0]['Count']
        Expected:
            831
        Got:
            np.int64(831)

### **REVIEW:** `pandas.pivot_table`

Suppose we want to build a table showing the total number of babies born of each sex in each year. One way is to `groupby` using both columns of interest:

In [97]:
babynames.groupby(["Year", "Sex"])[["Count"]].agg(sum).head(6)

/tmp/ipykernel_50798/3400362585.py:1: FutureWarning: The provided callable <built-in function sum> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  babynames.groupby(["Year", "Sex"])[["Count"]].agg(sum).head(6)


Count
Year Sex       
1910 F     5950
     M     3213
1911 F     6602
     M     3381
1912 F     9804
     M     8142

While this does give us the information we're looking for, a more natural approach is to use pivot tables to represent our data in a more readable format.

In [98]:
babynames_pivot = babynames.pivot_table(
    index = "Year",     # rows (turned into index)
    columns = "Sex",    # column values
    values = ["Count"], # field(s) to process in each group
    aggfunc = np.sum,   # group operation
)
babynames_pivot.head(6)

/tmp/ipykernel_50798/1884252639.py:1: FutureWarning: The provided callable <function sum at 0x7ba00a5b6840> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  babynames_pivot = babynames.pivot_table(


Count       
Sex       F      M
Year              
1910   5950   3213
1911   6602   3381
1912   9804   8142
1913  11860  10234
1914  13815  13111
1915  18643  17192

We can also include multiple values in our pivot tables 

In [99]:
babynames_pivot = babynames.pivot_table(
    index = "Year",     # rows (turned into index)
    columns = "Sex",    # column values
    values = ["Count", "Name"],
    aggfunc = np.max,   # group operation
)
babynames_pivot.head(6)

/tmp/ipykernel_50798/1132892114.py:1: FutureWarning: The provided callable <function max at 0x7ba00a5b7240> is currently using DataFrameGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  babynames_pivot = babynames.pivot_table(


Count          Name         
Sex      F     M       F        M
Year                             
1910   295   237  Yvonne  William
1911   390   214   Zelma   Willis
1912   534   501  Yvonne  Woodrow
1913   584   614   Zelma   Yoshio
1914   773   769   Zelma   Yoshio
1915   998  1033    Zita    Yukio

<br>

---

### Question 4

Using `presidential_candidates_and_name_popularity`, create a table, `party_result_popular_vote_pivot`, whose index is the `Party` and whose columns are their `Result`. Each cell should contain the total number of popular votes received. `pandas`' `pivot_table` documentation is [here](https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html).

You may notice that there are `NaN`s in your table from missing data. Replace the `NaN` values with 0. You may find `.pivot_table`'s `fill_value=` argument helpful. Or, you can use `pd.DataFrame.fillna` [(documentation here)](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html).

In [104]:
party_result_popular_vote_pivot = presidential_candidates_and_name_popularity.pivot_table(
    index = "Party",
    columns = "Result",
    values = "Popular vote",
    aggfunc = np.sum,
    fill_value = 0
)
party_result_popular_vote_pivot

/tmp/ipykernel_50798/2243758045.py:1: FutureWarning: The provided callable <function sum at 0x7ba00a5b6840> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  party_result_popular_vote_pivot = presidential_candidates_and_name_popularity.pivot_table(


Result                      loss        win
Party                                      
American                  158271          0
American Independent    11172260          0
Anti-Masonic              100715          0
Anti-Monopoly             134294          0
Citizens                  233052          0
Communist                 103307          0
Constitution              346721          0
Constitutional Union      590901          0
Democratic             314190254  441751531
Democratic-Republican     151271     113142
Free Soil                 446711          0
Green                    4254943          0
Greenback                 308649          0
Independent             28140589          0
Libertarian              7723416          0
National Democratic       134645          0
National Republican       985102          0
National Union                 0    2211317
Natural Law               113670          0
New Alliance              217221          0
Northern Democratic      1380202          0
Populist                 1261250          0
Progressive             10252501          0
Prohibition              2141790          0
Reform                   8085294          0
Republican             341121121  481561200
Socialist                4825871          0
Southern Democratic       848019          0
Taxpayers                 184656          0
Union                     892378          0
Whig                     1996929    2635818

In [105]:
grader.check("q4")

q4 results:
    q4 - 1 result:
        ✅ Test case passed

    q4 - 2 result:
        ✅ Test case passed

    q4 - 3 result:
        ❌ Test case failed
        Trying:
            party_result_popular_vote_pivot.isnull().sum().sum()
        Expecting:
            0
        **********************************************************************
        Line 1, in q4 2
        Failed example:
            party_result_popular_vote_pivot.isnull().sum().sum()
        Expected:
            0
        Got:
            np.int64(0)

    q4 - 4 result:
        ❌ Test case failed
        Trying:
            party_result_popular_vote_pivot.loc['American', 'Popular vote']['loss'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['American']['loss']
        Expecting:
            158271
        **********************************************************************
        Line 1, in q4 3
        Failed example:
            party_result_popular_vote_pivot.loc['American', 'Popular vote']['loss'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['American']['loss']
        Expected:
            158271
        Got:
            np.int64(158271)

    q4 - 5 result:
        ❌ Test case failed
        Trying:
            party_result_popular_vote_pivot.loc['American', 'Popular vote']['win'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['American']['win']
        Expecting:
            0
        **********************************************************************
        Line 1, in q4 4
        Failed example:
            party_result_popular_vote_pivot.loc['American', 'Popular vote']['win'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['American']['win']
        Expected:
            0
        Got:
            np.int64(0)

    q4 - 6 result:
        ❌ Test case failed
        Trying:
            party_result_popular_vote_pivot.loc['Democratic', 'Popular vote']['loss'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['Democratic']['loss']
        Expecting:
            314190254
        **********************************************************************
        Line 1, in q4 5
        Failed example:
            party_result_popular_vote_pivot.loc['Democratic', 'Popular vote']['loss'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['Democratic']['loss']
        Expected:
            314190254
        Got:
            np.int64(314190254)

    q4 - 7 result:
        ❌ Test case failed
        Trying:
            party_result_popular_vote_pivot.loc['Democratic', 'Popular vote']['win'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['Democratic']['win']
        Expecting:
            441751531
        **********************************************************************
        Line 1, in q4 6
        Failed example:
            party_result_popular_vote_pivot.loc['Democratic', 'Popular vote']['win'] if isinstance(party_result_popular_vote_pivot.columns, pd.MultiIndex) else party_result_popular_vote_pivot.loc['Democratic']['win']
        Expected:
            441751531
        Got:
            np.int64(441751531)

Just for fun: Which historical presidential candidates have names that were the least and most popular in 2022? Note: Here you'll observe a common problem in data science -- one of the least popular names is actually due to the fact that one recent president was so commonly known by his nickname that he appears named as such in the database from which you pulled election results.

In [112]:
# your optional code here
presidents = elections[elections['Result'] == 'win']
least_popular = presidents['Popular vote'].idxmin()
most_popular = presidents['Popular vote'].idxmax()
display(presidents.loc[least_popular])
display(presidents.loc[most_popular])

least_popular = presidents['%'].idxmin()
most_popular = presidents['%'].idxmax()
display(presidents.loc[least_popular])
display(presidents.loc[most_popular])

Year                             1824
Candidate           John Quincy Adams
Party           Democratic-Republican
Popular vote                   113142
Result                            win
%                           42.789878
Name: 1, dtype: object

Year                    2020
Candidate       Joseph Biden
Party             Democratic
Popular vote        81268924
Result                   win
%                  51.311515
Name: 178, dtype: object

Year                       1860
Candidate       Abraham Lincoln
Party                Republican
Popular vote            1855993
Result                      win
%                     39.699408
Name: 23, dtype: object

Year                      1964
Candidate       Lyndon Johnson
Party               Democratic
Popular vote          43127041
Result                     win
%                    61.344703
Name: 114, dtype: object

<br>

---

### Question 5
What if we want to use `groupby.agg` but aggregate using a different function for each column? One way to do this is by passing in a dictionary with the keys as column names and the values as function names to `.agg()`. 

Note: Shoutout to Connor for the question after lecture! 

#### Question 5a
Let's try it out. We are interested in learning a bit about our candidates' election experiences over the years. We want to know:

* the last year they ran,
* how many times they ran, and
* if they ever won.

Set `elections_multifn` to be a dataframe with this information, a result of calling `groupby.agg` using a dictionary of functions. `elections_multifn` should look similar to what is shown below. Small caveat: the `Party` column shown below may have a different column name.

<img src = "pics/elections_multifn.png" width = "300">

In [118]:
fn_dict = {'Year': 'last', 'Party': 'count', 'Result': lambda x: 'win' if (x == 'win').any() else 'loss'}
elections_multifn = elections.groupby("Candidate").agg(fn_dict)
elections_multifn

Year  Party Result
Candidate                                 
Aaron S. Watkins        1920      1   loss
Abraham Lincoln         1864      2    win
Adlai Stevenson         1956      2   loss
Al Gore                 2000      1   loss
Al Smith                1928      1   loss
...                      ...    ...    ...
William Z. Foster       1932      1   loss
Winfield Scott          1852      1   loss
Winfield Scott Hancock  1880      1   loss
Woodrow Wilson          1916      2    win
Zachary Taylor          1848      1    win

[132 rows x 3 columns]

In [119]:
grader.check("q5a")

q5a results: All test cases passed!

#### Question 5b

These column names are uninformative and incorrect. Rename the corresponding columns to be the following:
* `Most Recent Year`
* `Number of Elections`
* `Ever Won`

Hint: Remember you can use another dictionary here! 

In [120]:
col_dict = {'Year': 'Most Recent Year', 'Party': 'Number of Elections', 'Result': 'Ever Won'}
elections_multifn = elections_multifn.rename(columns = col_dict)
elections_multifn

Most Recent Year  Number of Elections Ever Won
Candidate                                                             
Aaron S. Watkins                    1920                    1     loss
Abraham Lincoln                     1864                    2      win
Adlai Stevenson                     1956                    2     loss
Al Gore                             2000                    1     loss
Al Smith                            1928                    1     loss
...                                  ...                  ...      ...
William Z. Foster                   1932                    1     loss
Winfield Scott                      1852                    1     loss
Winfield Scott Hancock              1880                    1     loss
Woodrow Wilson                      1916                    2      win
Zachary Taylor                      1848                    1      win

[132 rows x 3 columns]

In [121]:
grader.check("q5b")

q5b results: All test cases passed!

#### Question 5c

Now let's try to do it **without** using a dictionary of functions. First, create three dataframes, `elections_lastyr`, `elections_numtimes`, and `elections_everwon`, where each one is the result of applying the relevant function using `groupby.agg`. Then, merge the three of them using `pd.merge` twice. The final dataframe should be called `elections_multifn_manual`. 

Hint: You can either use `reset_index` to get `Candidate` back as a column or check out the `pd.merge` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.merge.html) to see how to merge on the index.

In [126]:
elections_lastyr = elections.groupby("Candidate").agg({'Year': 'last'})
elections_numtimes = elections.groupby("Candidate").agg({'Party': 'count'})
elections_everwon = elections.groupby("Candidate").agg({'Result': lambda x: 'win' if (x == 'win').any() else 'loss'})

elections_multifn_manual = pd.merge(elections_lastyr, elections_numtimes, on = 'Candidate')
elections_multifn_manual = pd.merge(elections_multifn_manual, elections_everwon, on = 'Candidate')

elections_multifn_manual

Year  Party Result
Candidate                                 
Aaron S. Watkins        1920      1   loss
Abraham Lincoln         1864      2    win
Adlai Stevenson         1956      2   loss
Al Gore                 2000      1   loss
Al Smith                1928      1   loss
...                      ...    ...    ...
William Z. Foster       1932      1   loss
Winfield Scott          1852      1   loss
Winfield Scott Hancock  1880      1   loss
Woodrow Wilson          1916      2    win
Zachary Taylor          1848      1    win

[132 rows x 3 columns]

In [127]:
grader.check("q5c")

q5c results: All test cases passed!

To finish, let's take a look at which candidates ran the most times...

In [128]:
elections_multifn.sort_values('Number of Elections', ascending=False).head(10)

Most Recent Year  Number of Elections Ever Won
Candidate                                                             
Norman Thomas                       1948                    5     loss
Franklin Roosevelt                  1944                    4      win
Eugene V. Debs                      1920                    4     loss
Ralph Nader                         2008                    4     loss
William Jennings Bryan              1908                    3     loss
Martin Van Buren                    1848                    3      win
Grover Cleveland                    1892                    3      win
Andrew Jackson                      1832                    3      win
Richard Nixon                       1972                    3      win
William McKinley                    1900                    2      win

<hr style="border: 5px solid #003262;" />
<hr style="border: 1px solid #fdb515;" />

## Congratulations! You have finished Lab 03!

Mitski and Guppy congratulate you!

<img src = "pics/mitski.jpg" width = "300"> <img src = "pics/guppy.jpg" width = "400">

### Course Content Feedback

If you have any feedback about this assignment or about any of our other weekly, weekly assignments, lectures, or discussions, please fill out the [Course Content Feedback Form](https://forms.gle/owfPCGgnrju1xQEA9). Your input is valuable in helping us improve the quality and relevance of our content to better meet your needs and expectations!

### Submission Instructions

Below, you will see a cell. Running this cell will automatically generate a zip file with your autograded answers. Submit this file to the Lab 03 assignment on Gradescope. If you run into any issues when running this cell, feel free to check this [section](https://ds100.org/debugging-guide/autograder_gradescope/autograder_gradescope.html#why-does-grader.exportrun_teststrue-fail-if-all-previous-tests-passed) in the Data 100 Debugging Guide.

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

In [ ]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False, run_tests=True)